cdap-cli/src/main/java/co/cask/cdap/cli/util/AbstractAuthCommand.java

```diff
@@ -72,29 +69,21 @@ public void execute(Arguments arguments, PrintStream printStream) throws Excepti
   public abstract void perform(Arguments arguments, PrintStream printStream) throws Exception;
 
   protected ServiceId parseServiceId(Arguments arguments) {
-    String[] appVersionServiceId = arguments.get(ArgumentName.SERVICE.toString()).split("\\.");
-    if (appVersionServiceId.length < 2) {
-      throw new CommandInputError(this);
-    }
-
-    String appId = appVersionServiceId[0];
-    String serviceId = appVersionServiceId[appVersionServiceId.length - 1];
-    String appVersion = appVersionServiceId.length == 2 ? ApplicationId.DEFAULT_VERSION :
-      dotJoiner.join(Arrays.copyOfRange(appVersionServiceId, 1, appVersionServiceId.length - 1));
-    return cliConfig.getCurrentNamespace().app(appId, appVersion).service(serviceId);
+    ProgramId programId = parseProgramId(arguments, ElementType.SERVICE);
+    return programId.getParent().service(programId.getProgram());
   }
 
   protected ProgramId parseProgramId(Arguments arguments, ElementType elementType) {
-    String[] programIdParts = arguments.get(elementType.getArgumentName().toString()).split("\\.");
-    if (programIdParts.length < 2) {
+    String[] argumentParts = arguments.get(elementType.getArgumentName().getName()).split("\\.");
+    if (argumentParts.length < 2) {
       throw new CommandInputError(this);
     }
 
-    String appId = programIdParts[0];
-    String programName = programIdParts[programIdParts.length - 1];
-    String appVersion = programIdParts.length == 2 ? ApplicationId.DEFAULT_VERSION :
-      dotJoiner.join(Arrays.copyOfRange(programIdParts, 1, programIdParts.length - 1));
-    return cliConfig.getCurrentNamespace().app(appId, appVersion).program(elementType.getProgramType(), programName);
+    String appName = argumentParts[0];
```

Comment:
I had a comment about creating a POJO for argument parts to make the code more readable. Can you check if that helps?